synapse/federation/sender/__init__.py

```diff
@@ -337,34 +341,84 @@ async def handle_event(event: EventBase) -> None:
                     logger.debug("Sending %s to %r", event, destinations)
 
                     if destinations:
-                        await self._send_pdu(event, destinations)
-
                         now = self.clock.time_msec()
                         ts = await self.store.get_received_ts(event.event_id)
 
                         synapse.metrics.event_processing_lag_by_event.labels(
                             "federation_sender"
                         ).observe((now - ts) / 1000)
 
-                async def handle_room_events(events: Iterable[EventBase]) -> None:
+                        return event, destinations
+                    return None
+
+                async def handle_room_events(
+                    room_id: str, events: Iterable[EventBase]
+                ) -> Tuple[str, Dict[EventBase, Collection[str]]]:
                     with Measure(self.clock, "handle_room_events"):
-                        for event in events:
-                            await handle_event(event)
+                        # Handle all events, skip if handle_event returns None
+                        handled_events = (
+                            ret
+                            for ret in [await handle_event(event) for event in events]
```

Comment:
if we change the `[]` comprehension into a generator (`()`), python will start throwing errors about async iterators, because to work with an async iterator in a comprehension, you have to prefix `for` with `async for`

This'll be a small memory tradeoff for readability, as this'd create a list of `len(events)`, but the surrounding generator will filter out all `None` values, which will have itself take up no space until it is consumed by the dict comprehension

so `len(events)` array of `Optional[Tuple[Eventbase, Collection[str]]`, then comprehended into `Dict[EventBase, Collection[str]]`